tools/testbench/topology.c

```diff
@@ -632,28 +632,26 @@ int load_process(void *dev, int comp_id, int pipeline_id,
 		printf("%u ", comp_ext.uuid[i]);
 	printf("\n");
 
-	/* Only one control is supported*/
-	if (widget->num_kcontrols > 1) {
-		fprintf(stderr, "error: more than one kcontrol defined\n");
-		return -EINVAL;
-	}
-
 	/* Get control into ctl and priv_data */
-	if (widget->num_kcontrols) {
+	for (i = 0; i < widget->num_kcontrols; i++) {
 		ret = tplg_load_one_control(&ctl, &priv_data, file);
 		if (ret < 0) {
 			fprintf(stderr, "error: failed control load\n");
 			return ret;
 		}
-	}
 
-	/* Merge process and priv_data into process_ipc */
-	ret = process_append_data(&process_ipc, &process, ctl, priv_data);
-	free(ctl);
-	free(priv_data);
-	if (ret) {
-		fprintf(stderr, "error: private data append failed\n");
-		return ret;
+		/* Merge process and priv_data into process_ipc */
+		if (priv_data)
+			ret = process_append_data(&process_ipc, &process, ctl, priv_data);
```

Comment:
unless `priv_data` is only allocated for one of controls, i.e. if there's one and only one `SND_SOC_TPLG_CTL_BYTES` control, you'll be overwriting `process_ipc` here and leaking memory? And below you're using `process_ipc` from the last iteration? Can `.num_kcontrols` be 0, then you'd be dereferencing `NULL` below?

Reply:
Good catch! I recall I had earlier problem here with multiple same type controls. Then tplg was changed into three different control types and I forgot about it. This this likely hits again, need to test.

Reply:
@lyakh I've hopefully now addressed your findings. As limitation this code can handle only one control private data append to new() IPC. A proper solution would need similar only cmd() based control passing to firmware. I feel it's too large task to do withing beamformer update and needs a separate PR.

This handler seems to work OK for testbench usage with the 5 controls presented in topology for the TDFB component.